Add ShapeShifterStamped class & synchronized throttle node

furushchev opened the pull request:
Add `synchronized_throttle` node which is almost the same as `topic_tools/throttle` but accepts multiple topics as inputs and timestamp of all output messages are synchronized.

This feature is useful when throttling both RGB and Depth image or stereo images used for 3dimensional processing.

For making the node work, I also added `topic_tools/ShapeShifterStamped` class which is almost the same as `topic_tools/ShapeShifter` but has `header` field in its instance for comparing timestamps on message synchronization.

![synchronized throttle nodelet 1](https://user-images.githubusercontent.com/1901008/39695772-aa0d3e0c-5226-11e8-9277-a1e40b999fa3.jpg)

dirk-thomas commented on tools/topic_tools/package.xml:
```diff
@@ -18,6 +18,8 @@
   <buildtool_depend version_gte="0.5.78">catkin</buildtool_depend>
 
   <build_depend>cpp_common</build_depend>
+  <build_depend>geometry_msgs</build_depend>
```
This would introduce a circular dependendcy between this repo and the `common_msgs` repo. That would make the first time release of these repositories problematic since both need each other.

Maybe moving this functionality into a separate repository would be better.

furushchev replied:
@dirk-thomas Thank you for the feedback! The root problem is there is no message which has Header field in this repo which is need for testing. Would it be nice if I create a new package (let the name be `test_topic_tools`) in [test](https://github.com/ros/ros_comm/tree/melodic-devel/test) directory and move all tests for this package into `test_topic_tools` and also generate a dummy message with header field?

dirk-thomas replied:
Sounds good to me :+1:

dirk-thomas replied:
@furushchev Friendly ping.

furushchev replied:
@dirk-thomas Thanks for pinging. Just now updated this PR according to the discussion above.

furushchev replied:
@dirk-thomas If it would be better to move all existing test codes in `topic_tools` to `test_topic_tools`, I'll do it.

dirk-thomas committed:
update test to pass with old and new yaml (#1893)

dirk-thomas committed:
fix flakyness of transform test (#1890)

dirk-thomas committed:
fix dictionary changed size during iteration (#1894)

dirk-thomas committed:
fix flakyness of test_topic_statistics (#1896)

* add assertion to narrow down error case

* update test to ignore messages which are insufficient

dirk-thomas committed:
increase time limit of advertisetest/publishtest.test to reduce flakyness (#1897)

* increase time limit of publishtest.test to reduce flakyness

* increase time limit of advertisetest.test to reduce flakyness

dirk-thomas committed:
make statistics.test less flaky (#1899)

mm318 committed:
fix AttributeError: __exit__ (#1915)

Signed-off-by: Miaofei <miaofei@amazon.com>

smnogar committed:
fix mac trying to use epoll instead of kqueue (#1907)

macOS was incorrectly attempting to use epoll/poll instead of kqueue, occassionally generating errors such as RuntimeError: concurrent poll() invocation and breaking python nodes.

Co-authored-by: Steve Nogar <snogar@gmail.com>

smnogar committed:
add exception for ConnectionAbortedError (#1908)

Prevent python nodes from spamming the terminal every time a node is quit. Noted in Python 3 running on macOS.

Co-authored-by: Steve Nogar <snogar@gmail.com>

dirk-thomas committed:
Merge pull request #1917 from ros/backports_melodic

changes between 1.15.4 (Noetic) and 1.14.4 for backporting into Melodic

dirk-thomas committed:
update changelogs

dirk-thomas committed:
1.14.5

Commit:
add topic_tools/ShapeShifterStamped class

Commit:
add test_topic_tools package

Commit:
add topic_tools/synchronized_throttle node

Commit:
add test for topic_tools/synchronized_throttle

dirk-thomas commented:
I retargeted the `noetic-devel` branch. Please rebase your branch and resolve conflicts.

dirk-thomas commented:
Closing due to no further activity.
